pkg/auth/userinfo.go

```diff
@@ -0,0 +1,94 @@
+/*
+Copyright 2022 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	authv1 "k8s.io/api/authentication/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	"k8s.io/utils/strings/slices"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/pkg/features"
+	"github.com/oam-dev/kubevela/pkg/oam"
+	"github.com/oam-dev/kubevela/pkg/oam/util"
+)
+
+const (
+	groupSeparator = ","
+)
+
+// ContextWithUserInfo inject username & group from app annotations into context
+// If serviceAccount is set and username is empty, identity will user the serviceAccount
+func ContextWithUserInfo(ctx context.Context, app *v1beta1.Application) context.Context {
+	if app == nil {
+		return ctx
+	}
+	return request.WithUser(ctx, GetUserInfoInAnnotation(&app.ObjectMeta))
+}
+
+// SetUserInfoInAnnotation set username and group from userInfo into annotations
+// it will clear the existing service account annotation in avoid of permission leak
+func SetUserInfoInAnnotation(obj *metav1.ObjectMeta, userInfo authv1.UserInfo) {
+	if AuthenticationWithUser {
+		metav1.SetMetaDataAnnotation(obj, oam.AnnotationApplicationUsername, userInfo.Username)
+	}
+	re := regexp.MustCompile(strings.ReplaceAll(AuthenticationGroupPattern, "*", ".*"))
+	var groups []string
+	for _, group := range userInfo.Groups {
+		if re.MatchString(group) {
+			groups = append(groups, group)
+		}
+	}
+	metav1.SetMetaDataAnnotation(obj, oam.AnnotationApplicationGroup, strings.Join(groups, groupSeparator))
+	util.RemoveAnnotations(obj, []string{oam.AnnotationApplicationServiceAccountName})
```

Comment:
Here a critical change is added. @devholic 
I realize that when in the mode that User is not attached to the application annotation, if the ServiceAccount annotation is set by user, it could lead to the leak of permissions.
For example, one user X only have a small scope of permissions but it added a service account name in the application annotation, which contains larger scope of permissions, the application will work with more permissions it should not have.
So the change here means, when enabling authentication in KubeVela, the ServiceAccount annotation will not work anymore. But in this case, setting up `--authentication-with-user` and adding ServiceAccount token to the creation process of Application will also work.

Reply:
Thanks for pointing this out 👍 

I think returning an error instead of mutating the request is better since it can give more information to the users and prevent them from unexpected behaviors. (Anyway, mutating sounds reasonable to me 🙂)

---

Maybe, we can enhance this feature by letting users annotate `allowed users` on the ServiceAccount, and checking it in the validating webhook. WDYT? (This is not important for now, so we may discuss this in the other issue.)

Reply:
@devholic Thanks for the reply :)

> I think returning an error instead of mutating the request is better since it can give more information to the users and prevent them from unexpected behaviors. (Anyway, mutating sounds reasonable to me 🙂)

I agree with that. So it's like returning BadRequest error from the mutating webhook when it saw the ServiceAccount annotation exists and the authentication option enabled?  

> Maybe, we can enhance this feature by letting users annotate `allowed users` on the ServiceAccount, and checking it in the validating webhook. WDYT? (This is not important for now, so we may discuss this in the other issue.)

I'm a bit confused with the `allowed users`. Is it an option in the validating webhook configured by the system maintainer? Not sure if I'm understanding it correctly. If so, I think it is useful for the non-authentication impersonation scene (which is the case without authentication switched on and using the ServiceAccount impersonation).

Reply:
> I think returning an error instead of mutating the request is better since it can give more information to the users and prevent them from unexpected behaviors. (Anyway, mutating sounds reasonable to me 🙂)

I moved this logic in mutating handler to validate the service account annotation non-existence when authentication enabled.

> Maybe, we can enhance this feature by letting users annotate allowed users on the ServiceAccount, and checking it in the validating webhook. WDYT? (This is not important for now, so we may discuss this in the other issue.)

We can reserve this enhancement discussion in future PR. :)

Reply:
Sorry for the late reply 🙇 

> I'm a bit confused with the allowed users. Is it an option in the validating webhook configured by the system maintainer? Not sure if I'm understanding it correctly. If so, I think it is useful for the non-authentication impersonation scene (which is the case without authentication switched on and using the ServiceAccount impersonation).

My suggestion was that the system maintainer may specify allowed users for the ServiceAccount. So if the allowed user tries to edit ServiceAccount-bound Application, there is no need to remove ServiceAccount annotation from the Application.

> We can reserve this enhancement discussion in future PR. :)

👍

Reply:
I see. We can do it in further PR. 👍🏼